tests/integration.rs

```diff
@@ -798,6 +825,82 @@ mod tests {
+                    let listener =
+                        TcpListener::bind(&lisen_addr.as_str()).map_err(Error::Connection)?;
+                    retry = false;
+
+                    // We may want to set a timeout here
+                    match listener.accept() {
+                        Ok((mut stream, addr)) => {
+                            // Make sure the connection is from the expected 'guest_addr'
+                            if addr.ip() != std::net::IpAddr::from_str(expected_guest_addr).unwrap()
+                            {
+                                s = format!("Expecting the guest ip '{}' while being connected with ip '{}'",
+                                            expected_guest_addr, addr.ip());
+                                return Err(Error::WrongGuestAddr);
+                            }
+
+                            // Make sure the right message is to notify the guest VM is booted
+                            let mut data = String::new();
+                            stream
+                                .set_read_timeout(Some(std::time::Duration::new(
+                                    DEFAULT_HTTP_SERVER_TIMEOUT as u64,
+                                    0,
+                                )))
+                                .map_err(Error::SetReadTimeout)?;
+                            stream
+                                .read_to_string(&mut data)
+                                .map_err(Error::ReadToString)?;
+                            if data != DEFAULT_HTTP_SERVER_MESSAGE {
+                                s = format!("Expecting the guest message '{}' while receiving the message '{}'",
+                                            DEFAULT_HTTP_SERVER_MESSAGE, data);
+                                return Err(Error::WrongGuestMsg);
+                            };
+
+                            Ok(())
+                        }
+                        Err(e) => {
+                            s = "TcpListener::accept() failed".to_string();
+                            Err(Error::Connection(e))
+                        }
+                    }
+                })() {
+                    Ok(_) => break,
+                    Err(e) => {
+                        counter += 1;
+                        if counter >= DEFAULT_HTTP_SERVER_RETRIES || !retry {
+                            eprintln!(
+                                "\n\n==== Start 'wait_vm_boot' (FAILED) ====\n\n\
+                                 lisen_addr=\"{}\"\n\
+                                 expected_guest_addr=\"{}\"\n\
+                                 message =\"{}\"\n\
+                                 \n==== End 'wait_vm_boot' outout ====\n\n",
+                                lisen_addr, expected_guest_addr, s,
+                            );
+
+                            return Err(e);
+                        }
+                    }
+                };
+                thread::sleep(std::time::Duration::new(
```

Comment:
This wasn't the way I expecting this to be fixed. Instead I was thinking that you would start the http server in another thread and have the main thread block on a condition variable (released after the http notification has been made.) Your solution looks like it should work but you may wait longer than necessary. Although your retry limit prevents it waiting for ever which would be slightly more faff with the thread + cond var approach.

Can you print out how long you had to wait and compare that  to the old static times?

Reply:
On my dev machine, the `wait_vm_boot` takes from 5.14s to 14.9s for `test_serial_*`. I also added a debugging commit that will print this information in our CI log.

Reply:
Okay, cool. So definitely an improvement. Could you tweak `DEFAULT_HTTP_SERVER_TIMEOUT` down to 5s?